### ISSUE TITLE
Does the container work with a wildcard ssl certificate?

### ISSUE THREAD

**Issue by Johnpojohn** [tomav/docker-mailserver#1338]
Im having some issues getting the container to work with a letsencrypt wildcard certificate (*.domain.com). My fqdn is mail.domain.com, and the mailserver container looks for a mail.domain.com folder at \etc\letsencrypt\live. However, if I mount a volume as directly:

\etc\letsencrypt\live\domain.com:\etc\letsencrypt\live\mail.domain.com

meaning it should have direct access to the cert and key, its still not working. Thus my question: Is the container supposed to work with a wildcard certificate at all, or do I have to request a certificate for mail.domain.com specifically? And if its possible, how should I mount it exactly?

Thanks!

**Comment by erik-wramner** [comment]
I'm not sure. There is an open pull request that sounds related, @fbartels do you know?

**Comment by fbartels** [comment]
@erik-wramner are you referring to https://github.com/tomav/docker-mailserver/pull/1301 ?

@Johnpojohn can you provide logs from your container startup? From that we should be able to see more clearer what is happening.

I don't think any of the services do care about the domain name that is encoded in the ssl certificate and the only way I can see this going wrong is when the filename the scripting expects is not present.

**Comment by Johnpojohn** [comment]
The logs look fine as soon as I mount the volume as described, the mailserver finds the cert files correctly (I was having some issues because apparently certbot only puts symlinks on the /live folder, but that's a separate problem). But I get invalid certificate complaints from all mail applications and if I check my TLS online, I get the following (personal references removed to *.domain.com):

Certificate 1 of 1 in chain: Cert VALIDATION ERROR(S): unable to get local issuer certificate; unable to verify the first certificate

Cert Hostname VERIFIED (mail.domain.com = *.domain.com | DNS:*.domain.com | DNS:domain.com)

**Comment by erik-wramner** [comment]
Yes, I meant #1301. At least that one works with the same parts of the code.

**Comment by fbartels** [comment]
> What I still get are invalid certificate complaints from all mail applications

Are you sure its using the right certificate? https://serverfault.com/a/661982

**Comment by Johnpojohn** [comment]
Fairly sure. A ssl cert check shows all correct parameters when checking for mail.domain.com (wildcard cert with common name *.domain.com, issued by letsencrypt). I only have one certificate for my whole system anyway. At this point I'm even wondering whether having wildcard certs for mailservers is at all viable.

**Comment by erik-wramner** [comment]
Both postfix and dovecot support wildcard certs, so that should not be a problem. Are there any intermediate certs? The entire chain must be present.

**Comment by smitpatelx** [comment]
I guess the same problem with me. Check this out https://github.com/tomav/docker-mailserver/issues/1341

**Comment by Johnpojohn** [comment]
> Both postfix and dovecot support wildcard certs, so that should not be a problem. Are there any intermediate certs? The entire chain must be present.

The entire chain is present. But this is indeed the same problem as #1341

**Comment by Johnpojohn** [comment]
I took a look at the solution of #1341 and for anyone having the exact same issue I do, it can be solved this way. The problem is twofold (assuming this is the only issue and the server works fine otherwise):

The first problem is that my docker cert container (linuxserver letsencrypt) does not place the cert files in /letsencrypt/live, but only simlinks. The actual files are in /letsencrypt/archive. The mailserver container is unable to work with these symlinks for whatever reason.

The second problem is that so far, I have not been able to make it work using the wildcard folder at /letsencrypt/live/domain.com. I don't know the reason for this, but it probably has something to do with the symlinks as well.

The solution is to create a new folder /letsencrypt/live/mail.domain.com, and manually copy all wildcard cert files from both /live/domain.com and /archive/domain.com (in this case renaming them from cert1.pem to cert.pem) to this folder. The mailserver container instructions to use letsencrypt certs will then work fine.

So in summary not really a problem of this container, but still quite a PITA if you want to get it to work.

Thanks for your help! And if anyone can suggest a way to fix it more permanently It'd be great.

**Comment by erik-wramner** [comment]
Good. Perhaps you could update the FAQ in the wiki to that effect? I'll close this then.

**Comment by polarathene** [comment]
> The first problem is that my docker cert container (linuxserver letsencrypt) does not place the cert files in /letsencrypt/live, but only symlinks. The actual files are in /letsencrypt/archive.
> The mailserver container is unable to work with these symlinks for whatever reason.
> Thanks for your help! And if anyone can suggest a way to fix it more permanently It'd be great.

@AugustoGarcia AFAIK, with a symlink it's not unlike say `localhost:8000/blog` would have little meaning to you(a different container) than what it does for me if running a service that responds to that. 

The symlink is saying "The actual file is *here*", but since you're only exposing the symlink, if the mailserver container is to try look at that same location, it will be checking a different filesystem/view than what the letsencrypt container has. (I could be wrong about this, I haven't verified)

If the letsencrypt container does not support using a volume or bind mount for that real location, then it'd be difficult to expose it to another container. It's possible to reverse bind mount, but that can be a bit of a hacky workaround. 

You could also look into an alternative service for handling letsencrypt, `acme.sh` for example has a Docker container that can run as a daemon to provision and renew your certs, while also being able to transfer / update certs to another container and run a command to restart any services within that container if needed.
